db/indexes.go

```diff
@@ -376,61 +376,89 @@ type InitializeIndexOptions struct {
+	fullIndexName string
+	sgIndex       SGIndex
+}
+
+type sgIndexesToBuild []sgIndexToBuild
+
+func (s sgIndexesToBuild) FullIndexNames() []string {
+	names := make([]string, 0, len(s))
+	for _, sgIndex := range s {
+		names = append(names, sgIndex.fullIndexName)
+	}
+	return names
+}
+
 // Initializes Sync Gateway indexes for datastore.  Creates required indexes if not found, then waits for index readiness.
 func InitializeIndexes(ctx context.Context, n1QLStore base.N1QLStore, options InitializeIndexOptions) error {
 	if options.NumPartitions < 1 {
 		return fmt.Errorf("Invalid number of partitions specified: %d, needs to be greater than 0", options.NumPartitions)
 	}
-	base.InfofCtx(ctx, base.KeyAll, "Initializing indexes with numReplicas: %d...", options.NumReplicas)
 
-	// Create any indexes that aren't present
-	fullIndexNames := make([]string, 0)
+	requiredIndexes := make(sgIndexesToBuild, 0, len(sgIndexes))
 	for _, sgIndex := range sgIndexes {
-
 		if !sgIndex.shouldCreate(options) {
 			base.DebugfCtx(ctx, base.KeyAll, "Skipping index: %s ...", sgIndex.simpleName)
 			continue
 		}
-
 		fullIndexName := sgIndex.fullIndexName(options.UseXattrs, options.NumPartitions)
+		requiredIndexes = append(requiredIndexes, sgIndexToBuild{fullIndexName: fullIndexName, sgIndex: sgIndex})
+	}
 
-		err := sgIndex.createIfNeeded(ctx, n1QLStore, options)
-		if err != nil {
-			if !errors.Is(err, base.ErrIndexBackgroundRetry) {
-				return base.RedactErrorf("Unable to install index %s: %v", base.MD(sgIndex.simpleName), err)
-			}
-		}
+	if len(requiredIndexes) == 0 {
+		// not expected to get here - all our collections require at least one index ...
+		base.AssertfCtx(ctx, "No indexes to create - expected at least one index to be created, but none were found")
+		return nil
+	}
 
-		fullIndexNames = append(fullIndexNames, fullIndexName)
+	indexesMeta, err := base.GetIndexesMeta(ctx, n1QLStore, requiredIndexes.FullIndexNames())
+	if err != nil {
+		base.WarnfCtx(ctx, "Error getting indexes meta: %v - continuing to create", err)
 	}
-	// Issue BUILD INDEX for any deferred indexes.
-	if len(fullIndexNames) > 0 {
-		buildErr := base.BuildDeferredIndexes(ctx, n1QLStore, fullIndexNames)
-		if buildErr != nil {
-			base.InfofCtx(ctx, base.KeyQuery, "Error building deferred indexes.  Error: %v", buildErr)
-			return buildErr
+
+	// Happy-path optimization (for each collection):
+	// Drop indexes from the list if they're already online to avoid any per-index init cost.
+	requiredIndexes = slices.DeleteFunc(requiredIndexes, func(idx sgIndexToBuild) bool {
+		meta, exists := indexesMeta[idx.fullIndexName]
+		if exists && meta.State == base.IndexStateOnline {
+			base.DebugfCtx(ctx, base.KeyAll, "Index %s is already online", idx.fullIndexName)
+			// remove from list of indexes to create/wait for
+			return true
 		}
-	}
+		base.DebugfCtx(ctx, base.KeyAll, "Index %s is not online - will issue CREATE INDEX statement and wait for readiness", idx.fullIndexName)
```

Comment:
I think this statement could be deceptive and I'd be inclined to drop it - while we do CREATE INDEX IF NOT EXISTS, this could be confusing in the case that the index exists but is in the process of building.

It looks like we don't log create index anyway but we do log in `WaitForIndexesOnline` when it is ready and when it is not ready and I think that's sufficient.

Reply:
Yep that's fair - we already have enough logging in the existing codepath.

```suggestion
```